main, wire, blockchain, indexers, ffldb: Add pruning

kcalvinalvin opened the pull request:
Adds pruning to btcd.

### Change to ffldb:

1) Add `PruneBlocks()` to transaction interface. Prune blocks is a no-op when the total block file size is equal to or under the target size. If the total block files size is over the target size, oldest block files are deleted along with the block indexes that point to those block files. Returns the hashes of the deleted blocks.

### Change to wire:

1) Add SFNodeNetworkLimited.
2) Add SFNodeNetworkLimited to default services.

### Change to blockchain:

1) New `pruneTarget` param to indicate how much the block files should take up.l
2) Attempts to prune blocks, spend journals, and relevant indexes on every `connectBlock`

### Change to indexers:

1) Added PruneBlock to Manager and implemented PruneBlock to every indexer.

### Change to main:

1) Prune flag is defined. It limits the prune to 1536MiB as we need to guarantee 288 blocks to abide by the NODE_NETWORK_LIMITED rule set by BIP159.
2) --addrindex and --prune at the same time is not allowed.
3) --txindex and --prune at the same time is not allowed.

Fixes https://github.com/btcsuite/btcd/issues/1069

saubyk commented:
cc @halseth in case you're interested in reviewing this pr. Thanks

halseth commented on database/ffldb/db.go:
```diff
@@ -1669,6 +1669,94 @@ func (tx *transaction) writePendingAndCommit() error {
+		return nil, makeDbErr(database.ErrTxNotWritable, str, nil)
+	}
+
+	if targetSize < tx.db.store.maxBlockFileSize {
```
not 288*blocksize?

kcalvinalvin replied:
No because I wanted the pruning behavior to be similar to bitcoin core's where the user can set a prune size. For example, the user could want a prune size of 70GB instead of the smallest possible size.

kcalvinalvin replied:
That's true that we have to keep at least 2 days worth of blocks but that's not the check being done here. The check here is if the database can actually reach the `targetSize` that the caller asked for. Since each of the `.blk` files are 512MiB each, it's impossible to reach anything less than that. For example, if the caller asks for a target size of  256MiB, it'll error out here.

That's the only guarantee that the database makes. The caller should be careful to not ask for anything lower than what it'd need. That's enforced in `config.go`

halseth commented on database/interface.go:
```diff
@@ -389,6 +389,15 @@ type Tx interface {
+	// be returned (other implementation-specific errors are possible):
+	//   - ErrTxNotWritable if attempted against a read-only transaction
+	//   - ErrTxClosed if the transaction has already been closed
+	PruneBlocks(targetSize uint32) ([][]byte, error)
```
document return value. Chould be `[]chainhash.Hash`?

kcalvinalvin replied:
Yes it could be `[]chainhash.Hash`. I'll update the code and document the return value

halseth reviewed (changes requested):
Looks pretty good to me :)

kcalvinalvin commented on database/ffldb/blockio.go:
```diff
@@ -713,36 +719,57 @@ func (s *blockStore) handleRollback(oldBlockFileNum, oldBlockOffset uint32) {
+		fileNum = strings.TrimSuffix(fileNum, blockFileExtension)
+
+		// Turn the string into a number.
+		lastFile, err = strconv.Atoi(fileNum)
```
Good point... Seems like `filepath.Glob` will spit out things in the order the files appear on your filesystem (which was always sorted).

I'll put in a sort so it's guaranteed the fileNames will be sorted.

guggero commented on database/ffldb/driver_test.go:
```diff
@@ -253,6 +255,129 @@ func TestPersistence(t *testing.T) {
+				if dbErr, ok := err.(database.Error); !ok ||
+					dbErr.ErrorCode != database.ErrBlockNotFound {
+
+					return fmt.Errorf("Expected ErrBlockNotFound but got nil")
```
Might not be `nil` but an unexpected error, so include it in the message?

kcalvinalvin replied:
Addressed in the latest push

guggero commented on database/interface.go:
```diff
@@ -389,6 +389,21 @@ type Tx interface {
+	//   - ErrTxNotWritable if attempted against a read-only transaction
+	//   - ErrTxClosed if the transaction has already been closed
+	//
+	// NOTE: The data returned by this function is only valid during a
```
The actual comment on the implementation should be identical to this one IMO.

kcalvinalvin replied:
I think it's better to have the comments separate since the interface and the implementation are different. It's also how other comments are done as well in the existing code.

guggero commented on blockchain/chain.go:
```diff
@@ -665,6 +670,37 @@ func (b *BlockChain) connectBlock(node *blockNode, block *btcutil.Block,
+
+			// No need to try to delete the spend journal entries or the indexes
+			// if we haven't deleted any blocks.
+			if deletedHashes == nil {
```
Check `len()` instead of `== nil` to ignore underlying implementation assumptions?

kcalvinalvin replied:
Addressed in the latest push

guggero commented on blockchain/indexers/cfindex.go:
```diff
@@ -255,6 +255,37 @@ func (idx *CfIndex) DisconnectBlock(dbTx database.Tx, block *btcutil.Block,
 
+// PruneBlock is invoked when an older block is deleted after it's been
+// processed.  For cfindx, PruneBlock is exactly the same as disconnecting a block.
+// TODO (kcalvinalvin): Consider keeping the filters at a later date to help with
```
I guess we could also support fast rescan using compact filters in the future when pruning? Maybe it would be a great idea to have an option like `prune-cfilters=true/false` that is `false` by default (since filters for mainnet are currently only around 8 GB).

kcalvinalvin replied:
I just realized that the fast rescan for bitcoin core is for the wallet. Since btcd doesn't have any wallet functionality with itself, there isn't a need for the compact filters for this reason.

Does btcwallet ask for the filters? Might be worth keeping the filters around for it?

Roasbeef replied:
> Does btcwallet ask for the filters(not familiar with the codebase)? Might be worth keeping the filters around for it?

We don't do that _yet_ but it's something we've been meaning on adding.

guggero commented on wire/protocol.go:
```diff
@@ -93,18 +93,23 @@ const (
 	// software.
 	SFNode2X
+
+	// SFNodeNetWorkLimited is a flag used to indicate a peer supports serving
```
Ah, if this is an actual p2p requirement that we need to be able to serve at least the last 288 blocks, I think we should update our minimum prune value accordingly (see previous comment of not using the max block file size as the minimum).

kcalvinalvin replied:
Refer to the reply here: https://github.com/btcsuite/btcd/pull/1971#discussion_r1200443717

losh11 replied:
I believe `SFNodeNetworkLimited` is untyped here

guggero commented on config.go:
```diff
@@ -66,6 +66,7 @@ const (
 	sampleConfigFilename         = "sample-btcd.conf"
 	defaultTxIndex               = false
 	defaultAddrIndex             = false
+	pruneMinSize                 = 1536
```
Ah, I see there's another min size here. Shouldn't we sync this with what's being checked in `database/ffldb/db.go`?

kcalvinalvin replied:
So the 1536 number came from the fact that we have these constants.

```
blocksNeeded = 288
ConsensusMaxBlockSize = 4,000,000
onDiskBlkFileSizeInBytes = 536,870,912
```

We need to keep 288 blocks to adhere to `NODE_NETWORK_LIMITED` BIP. Each block can be a maximum of 4MB. Each block file (XXXXXX.fdb) is 512MiB.

```
(blocksNeeded*ConsensusMaxBlockSize)/onDiskBlkFileSizeInBytes = 2.145
```

We do the math and we need to keep 2.14 on disk .fdb files to adhere to the `NODE_NETWORK_LIMITED` rule. Can't keep .14 files so we round up to 3. `512MiB * 3` is 1536MiB which is out `pruneMinSize`.

As mentioned [here](https://github.com/btcsuite/btcd/pull/1971#discussion_r1211140556), the check in `database/ffldb/db.go` is a different check so it's ok.

ProofOfKeags replied:
Yeah I don't mind the intervals at all. It's hardly an issue now that the overall prune cache size isn't maxed at ~4G

guggero replied:
Ah okay, the two different limits now make sense to me, thanks for the explanation.

guggero commented on config.go:
```diff
@@ -145,6 +146,7 @@ type config struct {
 	Proxy                string        `long:"proxy" description:"Connect via SOCKS5 proxy (eg. 127.0.0.1:9050)"`
 	ProxyPass            string        `long:"proxypass" default-mask:"-" description:"Password for proxy server"`
 	ProxyUser            string        `long:"proxyuser" description:"Username for proxy server"`
+	Prune                uint32        `long:"prune" description:"Prune already validated blocks from the database. Must specify a target size in MiB (minimum value of 1024)"`
```
Description doesn't match chosen min size constant value.

kcalvinalvin replied:
Addressed in the latest push

guggero reviewed (commented):
Great work, very cool to see such an impactful feature being introduced in only 8 commits! Did a first pass review, will do some manual testing in the next round.

kcalvinalvin commented on database/ffldb/db.go:
```diff
@@ -1669,6 +1669,94 @@ func (tx *transaction) writePendingAndCommit() error {
+
+		_, found := deletedFiles[loc.blockFileNum]
+		if found {
+			deletedBlockHashes = append(deletedBlockHashes, *(*chainhash.Hash)(cursor.Key()))
```
Not quite sure what you mean here. Mine is set to 8 spaces already and I couldn't find examples of what you mentioned here.

Also, afaik, tabspace shouldn't matter with tabs anyways since the editor you're using should render it as whatever you set right?

guggero replied:
I mean that the code in this PR often flows over the targeted 80 character line length limit we use in `lnd`. Though I see that is not specifically mentioned in this project's code contribution guidelines. So feel free to ignore as much of the existing code also doesn't seem to use that limit.

kcalvinalvin commented on database/ffldb/blockio.go:
```diff
@@ -713,36 +720,58 @@ func (s *blockStore) handleRollback(oldBlockFileNum, oldBlockOffset uint32) {
+// position at the last file is considered the current write cursor which is
+// also stored in the metadata.  Thus, it is used to detect unexpected shutdowns
+// in the middle of writes so the block files can be reconciled.
+func scanBlockFiles(dbPath string) (firstFile int, lastFile int, lastFileLen uint32, err error) {
```
Changed to `return firstFile, lastFile, lastFileLen, nil` in the latest push

kcalvinalvin commented on database/ffldb/blockio.go:
```diff
@@ -713,36 +720,58 @@ func (s *blockStore) handleRollback(oldBlockFileNum, oldBlockOffset uint32) {
+	for _, file := range files {
+		// Extract the file number.
+		fileNum := file
+		fileNum = strings.TrimPrefix(fileNum, dbPath+"/")
```
Addressed in the latest push

kcalvinalvin commented on database/ffldb/blockio.go:
```diff
@@ -713,36 +720,58 @@ func (s *blockStore) handleRollback(oldBlockFileNum, oldBlockOffset uint32) {
+func scanBlockFiles(dbPath string) (firstFile int, lastFile int, lastFileLen uint32, err error) {
+	// The only error we can get is a bad pattern error.  Since we're hardcoding
+	// the pattern, we should not have an error at runtime.
+	files, _ := (filepath.Glob(filepath.Join(dbPath, "*"+blockFileExtension)))
```
Addressed in the latest push

kcalvinalvin commented on database/ffldb/db.go:
```diff
@@ -1669,6 +1669,94 @@ func (tx *transaction) writePendingAndCommit() error {
 }
 
+// PruneBlocks deletes the block files until it reaches the target size
+// (specificed in bytes).  Throws an error if the target size is below
```
Addressed in the latest push

kcalvinalvin commented on database/ffldb/db.go:
```diff
@@ -1669,6 +1669,94 @@ func (tx *transaction) writePendingAndCommit() error {
+
+	// If the total size of block files are under the target, return early and
+	// don't prune.
+	totalSize := lastFileSize + (tx.db.store.maxBlockFileSize * uint32(maxSizeFileCount))
```
Addressed in the latest push

kcalvinalvin commented on database/ffldb/db.go:
```diff
@@ -1669,6 +1669,94 @@ func (tx *transaction) writePendingAndCommit() error {
+		targetSize/(1024*1024))
+
+	deletedFiles := make(map[uint32]struct{})
+	for i := uint32(first); i <= uint32(last); i++ {
```
Addressed in the latest push. There were other checks to stop it from happening but change it anyways since it doesn't hurt

kcalvinalvin commented on database/ffldb/driver_test.go:
```diff
@@ -253,6 +255,129 @@ func TestPersistence(t *testing.T) {
+						"%v", i, err)
+				}
+			}
+			return err
```
Addressed in the latest push

kcalvinalvin commented on database/ffldb/db.go:
```diff
@@ -1669,6 +1669,94 @@ func (tx *transaction) writePendingAndCommit() error {
+	// don't prune.
+	totalSize := lastFileSize + (tx.db.store.maxBlockFileSize * uint32(maxSizeFileCount))
+	if totalSize <= targetSize {
+		return nil, err
```
Addressed in the latest push

kcalvinalvin commented on database/ffldb/db.go:
```diff
@@ -1669,6 +1669,94 @@ func (tx *transaction) writePendingAndCommit() error {
+
+	// If we have no files on disk or just a single file on disk, return early.
+	if first == last {
+		return nil, err
```
Addressed in the latest push

kcalvinalvin commented on database/ffldb/driver_test.go:
```diff
@@ -253,6 +255,129 @@ func TestPersistence(t *testing.T) {
+	dbPath := filepath.Join(os.TempDir(), "ffldb-prunetest")
+	_ = os.RemoveAll(dbPath)
+	db, err := database.Create(dbType, dbPath, blockDataNet)
+	if err != nil {
```
This one? https://pkg.go.dev/github.com/stretchr/testify/require

I don't see it in the go.mod though

guggero replied:
It's here: https://github.com/btcsuite/btcd/blob/3faf68c8c57d925baea2bdcbcbc8a7efbbf99080/go.mod#L16, we mainly started using it in the `txscript` and `btcec` packages. It reduces most assertions to just a single line which I think is nice. But non blocking.

kcalvinalvin commented on database/ffldb/driver_test.go:
```diff
@@ -253,6 +255,129 @@ func TestPersistence(t *testing.T) {
+
+	// Create a new database to run tests against.
+	dbPath := filepath.Join(os.TempDir(), "ffldb-prunetest")
+	_ = os.RemoveAll(dbPath)
```
Addressed in the latest push

kcalvinalvin commented on database/ffldb/blockio.go:
```diff
@@ -713,36 +720,58 @@ func (s *blockStore) handleRollback(oldBlockFileNum, oldBlockOffset uint32) {
+func scanBlockFiles(dbPath string) (firstFile int, lastFile int, lastFileLen uint32, err error) {
+	// The only error we can get is a bad pattern error.  Since we're hardcoding
+	// the pattern, we should not have an error at runtime.
+	files, _ := (filepath.Glob(filepath.Join(dbPath, "*"+blockFileExtension)))
```
Addressed in the latest push

ProofOfKeags commented on blockchain/indexers/txindex.go:
```diff
@@ -430,6 +430,16 @@ func (idx *TxIndex) DisconnectBlock(dbTx database.Tx, block *btcutil.Block,
+//
+// This is part of the Indexer interface.
+func (idx *TxIndex) PruneBlock(dbTx database.Tx, blockHash chainhash.Hash) error {
+	return nil
```
seems like this ought to be an error given the rest of this PR

ProofOfKeags commented on blockchain/indexers/addrindex.go:
```diff
@@ -809,6 +809,16 @@ func (idx *AddrIndex) DisconnectBlock(dbTx database.Tx, block *btcutil.Block,
+//
+// This is part of the Indexer interface.
+func (idx *AddrIndex) PruneBlock(dbTx database.Tx, blockHash chainhash.Hash) error {
+	return nil
```
seems like this ought to be an error given the rest of this PR

ProofOfKeags commented on database/ffldb/blockio.go:
```diff
@@ -713,36 +720,61 @@ func (s *blockStore) handleRollback(oldBlockFileNum, oldBlockOffset uint32) {
+	lastFile, firstFile = -1, -1
+	for _, file := range files {
+		// Extract the file number.
+		fileNum := file
+		fileNum = filepath.Base(fileNum)
+		fileNum = strings.TrimSuffix(fileNum, blockFileExtension)
```
Inline this.

```suggestion
		fileNum := strings.TrimSuffix(filepath.Base(file), blockFileExtension)
```

If you don't want to inline it I strongly suggest not mutating `fileNum` like this and just create different names for the fragments. If you're worried about overallocation consider allocating stack space outside the loop and only setting each of these values once inside the loop.

ProofOfKeags commented on database/ffldb/blockio.go:
```diff
@@ -713,36 +720,61 @@ func (s *blockStore) handleRollback(oldBlockFileNum, oldBlockOffset uint32) {
+		// Actually get the file info.
+		filePath := blockFilePath(dbPath, uint32(lastFile))
 		st, err := os.Stat(filePath)
 		if err != nil {
-			break
+			return 0, 0, 0, err
+		}
+		if firstFile == -1 {
+			firstFile = lastFile
 		}
-		lastFile = i
 
-		fileLen = uint32(st.Size())
+		lastFileLen = uint32(st.Size())
```
Move this outside the loop. This is making a bunch of syscalls where we immediately throw away the result.

ProofOfKeags reviewed (changes requested):
There are a couple of things I think are necessary to change like the Prune Cache uint width. There are a number of things that seem fishy to me architecturally but nothing I can point to that's broken about it. I think even with the architectural smells that the benefit of a prunable btcd is worth merging these changes.

kcalvinalvin commented on blockchain/chain.go:
```diff
@@ -114,6 +115,10 @@ type BlockChain struct {
 
+	// pruneTarget is the size in bytes the database targets for when the node
+	// is pruned.
+	pruneTarget uint32
```
Good catch! Addressed in the latest push

ProofOfKeags commented on config.go:
```diff
@@ -1136,6 +1138,30 @@ func loadConfig() (*config, []string, error) {
 		}
 	}
 
+	if cfg.Prune != 0 && cfg.Prune < pruneMinSize {
+		err := fmt.Errorf("%s: the minimum value for --prune is %d. Got %d",
+			funcName, pruneMinSize, cfg.Prune)
+		fmt.Fprintln(os.Stderr, err)
+		fmt.Fprintln(os.Stderr, usageMessage)
+		return nil, nil, err
+	}
```
Seems reasonable

guggero reviewed (commented):
Amazing work, I think the code looks good to go.

I tested several things on testnet and found no bugs:
 - Sync from empty state to chain tip without pruning
 - Reduce pruning to large value slightly below the actual size (30GB), expect only a few block files to be pruned
 - Reduce pruning to 20GB, expect more block files to be pruned
 - Reduce pruning to min value, expect all but three block files to be pruned
 - Remove pruning again, make sure the node continues working
 - Start new sync from empty state to chain tip with pruning set to minimal value, expect there never to be more than 3 block files to be present


I found one small UX thing that we might want to fix:
When running `btcctl getblockchaininfo` there's a boolean `pruned` that is always set to `false`, even when pruning is turned on. Should be fairly easy to add, so I'm waiting for that before fully approving the PR.

Roasbeef commented on database/ffldb/blockio.go:
```diff
@@ -713,36 +720,61 @@ func (s *blockStore) handleRollback(oldBlockFileNum, oldBlockOffset uint32) {
+// also stored in the metadata.  Thus, it is used to detect unexpected shutdowns
+// in the middle of writes so the block files can be reconciled.
+func scanBlockFiles(dbPath string) (
+	firstFile int, lastFile int, lastFileLen uint32, err error) {
```
So we typically never used named returns, as for the most part they make code more difficult to follow in line. So in this case, we'd just declare them upfront, or even return a new struct with the field unfolded.

kcalvinalvin replied:
Addressed in the latest push.

Roasbeef commented on database/ffldb/blockio.go:
```diff
@@ -713,36 +720,61 @@ func (s *blockStore) handleRollback(oldBlockFileNum, oldBlockOffset uint32) {
+	sort.Strings(files)
+
+	lastFile, firstFile = -1, -1
+	for i, file := range files {
```
Why do we need to actually _scan_ all the files here? With this new usage of `filepath.Glob` we get all the file names up front, then sort them. So we can just get the first, and the last file directly using indexes. Then this way, we don't need to continue scanning and stat'ing them all after we've found the first and last file above.

kcalvinalvin replied:
One thing I had in mind was that you could technically have a `511MiB` `.fdb` file if the last block being written onto the disk doesn't fit.

But yeah you'd save a lot of i/o by just skipping. I'll implement that since that seems better overall.

Roasbeef commented on database/ffldb/db.go:
```diff
@@ -1669,6 +1669,99 @@ func (tx *transaction) writePendingAndCommit() error {
+	}
+
+	log.Tracef("Using %d more bytes than the target of %d MiB. Pruning files...",
+		totalSize-targetSize,
```
As is, doesn't this fail to account for the size of the last file on disk which may not be full? Or is the intended feature contract here that: we support pruning, but you can have `N GB + maxBlockFile` bytes on disk?

kcalvinalvin replied:
`N GB + maxBlockFile` is the current implementation.

I didn't give that much thought into it since you'll at max be `512MiB` off. Which wouldn't really change the user experience as the utxo set is much bigger than that.

Since we're already fetching the `lastFileSize`, I could change it so that the last block file length is accounted for in the size limit.

Roasbeef commented on database/ffldb/db.go:
```diff
@@ -1669,6 +1669,99 @@ func (tx *transaction) writePendingAndCommit() error {
+
+	deletedFiles := make(map[uint32]struct{})
+
+	// < not <= so that the last file is never deleted.  There are other checks in place
```
Style nit: can prefix with "We use" so it becomes a complete sentence.

Roasbeef commented on database/ffldb/db.go:
```diff
@@ -1669,6 +1669,99 @@ func (tx *transaction) writePendingAndCommit() error {
+	for ok := cursor.First(); ok; ok = cursor.Next() {
+		loc := deserializeBlockLoc(cursor.Value())
+
+		_, found := deletedFiles[loc.blockFileNum]
```
Why would there ever be a gap here? Shouldn't we always be deleting contiguous block files?

kcalvinalvin replied:
Is the order of blocks being saved onto the disk guaranteed?

Here we're looping over the blockLocations that are mapped by the blockhash, not the actual blocks in the `*.fdb` filess (that happens in the above loop). You can't assume which blocklocation was in which `*.fdb` file so I loop through them and delete them individually.

Roasbeef commented on database/ffldb/driver_test.go:
```diff
@@ -253,6 +255,127 @@ func TestPersistence(t *testing.T) {
+				if dbErr, ok := err.(database.Error); !ok ||
+					dbErr.ErrorCode != database.ErrBlockNotFound {
+
+					return fmt.Errorf("Expected ErrBlockNotFound but got %v", dbErr)
```
Overflows 80 char columns, a few other instances of this in the diff.

kcalvinalvin replied:
My personal preference is that 80 char limit doesn't really make sense in the current world. Even the linux kernel now does a default of 100 chars.

Since the rest of the codebase isn't doing 80 char either, does it make sense to enforce this here? I can wrap it here but I do think that the rule is outdated.

Roasbeef replied:
Ah yeah you're right this this didn't make it into the initial version of our [code contribution guidelines](https://github.com/btcsuite/btcd/blob/master/docs/code_contribution_guidelines.md) but it _has_ been adhered to as a standard nevertheless. I'll seek to have it updated so we can also use the linter to flag instances automatically in the future.

kcalvinalvin replied:
Addressed in the latest push

Roasbeef commented on database/ffldb/driver_test.go:
```diff
@@ -253,6 +255,127 @@ func TestPersistence(t *testing.T) {
+		}
+
+		var deletedBlocks []chainhash.Hash
+		// This should leave 3 files on disk.
```
Style nit: should have a newline above the comment.

kcalvinalvin replied:
Addressed in the latest push

Roasbeef commented on database/ffldb/driver_test.go:
```diff
@@ -253,6 +255,127 @@ func TestPersistence(t *testing.T) {
+		}
+
+		err = db.Update(func(tx database.Tx) error {
+			_, err := tx.PruneBlocks(1024)
```
What about the boundary cases when: try to prune more than allowed, no files on disk, already pruned to target, etc.

kcalvinalvin replied:
The current test will test for `try to prune more than allowed`.

I'm not sure if `no files on disk` and `already pruned to target` should end in an error? For both cases, it's meeting the requirement of: `sizeOnDisk < targetPruneSize`.

Roasbeef commented on blockchain/chain.go:
```diff
@@ -667,6 +671,37 @@ func (b *BlockChain) connectBlock(node *blockNode, block *btcutil.Block,
+		err = b.db.Update(func(dbTx database.Tx) error {
+			// When the total block size is under the prune target, prune blocks is
+			// a no-op and the deleted hashes are nil.
+			deletedHashes, err := dbTx.PruneBlocks(b.pruneTarget)
```
Why not run this in the above transaction where all the other indexes are updated?

Roasbeef commented on blockchain/indexers/addrindex.go:
```diff
@@ -809,6 +809,16 @@ func (idx *AddrIndex) DisconnectBlock(dbTx database.Tx, block *btcutil.Block,
 
+// PruneBlock is invoked when an older block is deleted after it's been
+// processed.
+// NOTE: For AddrIndex, it's a no-op as AddrIndex isn't allowed to be enabled
```
Wouldn't we want to _reclaim_ the space on disk due to the addr index? Or is the idea that you need to first drop the entire thing, _then_ enable pruning?

kcalvinalvin replied:
Currently if you have had the addr index enabled, it won't prune the ones you've already indexed unless you do it manually.

I guess we could do either one of these:

1: Don't do anything with the already present addr index data (current)
2: Drop the addr index if pruning is enabled
3: Force the user to drop the addr index before allowing them to enable pruning

I'd favor either 1 or 3. Explicit user action or doing nothing seems better.

Roasbeef replied:
> 3: Force the user to drop the addr index before allowing them to enable pruning

I prefer this path myself. As otherwise, the queries of indexes like the addr index wouldn't be possible as it wants to return the block regions where the transaction exist, but they may not be accessible after pruning. 

The only index that can still serve its purpose with pruned data is the neutrino filter index, as querying it doesn't require any block data at all (just slinging out the premade filters for each block).

Roasbeef commented on blockchain/indexers/cfindex.go:
```diff
@@ -255,6 +255,37 @@ func (idx *CfIndex) DisconnectBlock(dbTx database.Tx, block *btcutil.Block,
+//
+// This is part of the Indexer interface.
+func (idx *CfIndex) PruneBlock(dbTx database.Tx, blockHash chainhash.Hash) error {
+	for _, key := range cfIndexKeys {
```
We can do a small refactor here to allow this inner function to take the block hash directly, then we can call it for `DisconnectBlock` as well.

Roasbeef commented on server.go:
```diff
@@ -44,8 +44,8 @@ import (
-	defaultServices = wire.SFNodeNetwork | wire.SFNodeBloom |
-		wire.SFNodeWitness | wire.SFNodeCF
+	defaultServices = wire.SFNodeNetwork | wire.SFNodeNetworkLimited |
+		wire.SFNodeBloom | wire.SFNodeWitness | wire.SFNodeCF
```
If pruning isn't active, then this bit should be removed in `newServer`: 
https://github.com/btcsuite/btcd/blob/85687d455d8cca1077dae5fbf20accd59cb00da9/server.go#L2726-L2733

kcalvinalvin replied:
In Core, `NODE_NETWORK_LIMITED` is always on and `NODE_NETWORK` bit is added if the node isn't pruned.

https://github.com/bitcoin/bitcoin/blob/a4ca4975880c4f870c6047065c70610af2529e74/src/init.cpp#L786

https://github.com/bitcoin/bitcoin/blob/a4ca4975880c4f870c6047065c70610af2529e74/src/init.cpp#L1598C13-L1601

This is reflected in bitnodes.io for archival nodes https://bitnodes.io/nodes/18.27.79.17-8333/

I think it's better to follow the behavior of Core nodes since it's what everyone expects.

Roasbeef reviewed (changes requested):
Nice work! I love how small the final diff actually is. I had prior versions that attempted to do more fine grained accounting, which weren't anywhere near as clean as this implementation is. 

My main question in the diff is how we should handle the existing indexes already preseent:
  * Should we still allow `txindex` since it allows you to fetch blocks beyond the prune horizon? 
  * Should we attempt to delete the old information for existing indexes, making them partially present, or just wipe the entire thing? 
  * Should we continue to keep the compact filters as well, since they can be used to still serve light clients?

kcalvinalvin commented:
> * Should we still allow `txindex` since it allows you to fetch blocks beyond the prune horizon?

I'm not sure. We could allow for it and just return empty strings when you attempt to fetch an already pruned block.

> * Should we attempt to delete the old information for existing indexes, making them partially present, or just wipe the entire thing?

Refer to https://github.com/btcsuite/btcd/pull/1971#discussion_r1284037346

> * Should we continue to keep the compact filters as well, since they can be used to still serve light clients?

I'm indifferent either way. The user does have the option to drop the entire thing if they want. Though it may be nice to be able to only keep a portion of it?

halseth reviewed (approved):
LGTM 🦩

guggero reviewed (commented):
> When running btcctl getblockchaininfo there's a boolean pruned that is always set to false, even when pruning is turned on. Should be fairly easy to add, so I'm waiting for that before fully approving the PR.

I think this isn't addressed yet. Other than that (and the other remaining comments), this is good to go!

Roasbeef commented on database/ffldb/db.go:
```diff
@@ -1669,6 +1669,99 @@ func (tx *transaction) writePendingAndCommit() error {
+			}
+		}
+	}
+
```
I think let's also log here the new size once the pruning has been completed? That way the logs can be used approximate how long the pruning step took: 
```
2023-08-15 17:24:25.031 [TRC] BCDB: Scan found latest block file #54 with length 215346498
2023-08-15 17:24:25.031 [TRC] BCDB: Using 23963495746 more bytes than the target of 5000 MiB. Pruning files...
2023-08-15 17:25:08.715 [DBG] CHAN: Accepted block 0000000000000008a838becbe29d3a29fe89e291b8ad18fdf9021f8024585b2d
```

kcalvinalvin commented:
Made some additional changes:

1: Correctly return the prune status for `getblockchaininfo`
2: Got rid of pruning code for indexers. Compact filters are no longer pruned when the blocks are deleted.
3: Force user to drop already existing addr and/or tx index when trying to enable pruning.
4: Error if user enables addr and/or tx index while pruned.
5: Error if user tries to enable cfindex if they started the ibd while pruned and cfindex disabled. 
5: Force user to drop already indexed cfindex data when trying to disable cfindex while pruned.

kcalvinalvin commented:
The last push just adds a more descriptive error message if the user tries to disable `cfindex` while the node is already pruned and has cfindex enabled.

kcalvinalvin commented:
Sorry for yet another additional push. 3 things here.

1) The biggest thing: I realized that a lot of the code depends on `cfg.Prune` to tell if the node is pruned or not. But the code allowed the user to omit the flag even if the node had been previously pruned. This would've resulted in broadcasting to peers that the node isn't pruned and also would've given the wrong prune status for `getblockchaininfo`.

Instead of trying to see every case where I'd also have to call the database for the prune status, I added a commit that forces the user to prune if the database had been pruned previously like bitcoind. I don't think this really affects the UX and negates future bugs as well that can arise from not checking db for the prune status and just relying on `cfg.Prune`.

2) `getblockchaininfo` now depends on `cfg.Prune != 0`
3) Added additional message for the `--prune` flag on `-h` that tells the user about the default value of `cfg.Prune == 0` and how that disables pruning.

Roasbeef commented:
>  This would've resulted in broadcasting to peers that the node isn't pruned

Can you elaborate on this? Broadcasting as in the set of services bits we advertise, or something else? IIUC, those service bits will always give away if a node has been pruned or not.

Roasbeef reviewed (approved):
LGTM 🎏

+1 to the additional safeguarding around user configuration/downgrades
